### PR TITLE
Fix: added defi opportunity card component

### LINKED ISSUE #955: DeFi Table Opportunity Redesign
## Description 📹

**We’ve added new cards that display the same parameters as the table. These 3 cards highlight the top yield opportunities based on the selected filters. By default, they show the 3 best opportunities overall. For example, if you filter by USDC token and lending, the cards will display the 3 best yield opportunities for lending USDC. So you can use the first 3 rows of the table based on the filters**

Current table : https://starknet.quest/discover/defi
Refont table : 
image

Design link: https://www.figma.com/design/fh0OAvj4AS08kHoSxu3DkE/%F0%9F%9A%80-Starknet-Quest?node-id=6331-10149&t=9Ky1zTkRkLF10uV9-1**

## Proposed Actions 🛠️

Here’s a checklist of actions to follow for resolving this issue:

1. **Fork and Create Branch**:  
   Fork the repository and create a new branch using the issue number:
   ```bash
   git checkout -b fix-[issue-number]
   ```

2. **Implement Changes**:  
- [ ] For this issue we'll only need you to add the cards highlighting the 3 best opportunities, you'll need you to wait for #954 to be done
- [ ] You can create a Card component without adding more dependencies and [use this architecture](https://v0.dev/chat/B0eTJshQO61?b=b_A7jicLELGGa&p=0) 
- [ ] Then place your component in the `defiTable.tsx` component like so

```tsx
            <div className="w-fit  modified-cursor-pointer border-[1px] border-[#f4faff4d] px-4 rounded-md">
              <Typography type={TEXT_TYPE.BODY_DEFAULT} color="white">
                Clear All
              </Typography>
            </div>
          </div>
        </div>
        {/* Cards HERE */}
        <div className="rounded-xl border-[1px] border-[#f4faff4d] min-w-[930px] xl:w-full">
          <Table>
            <TableHeader>
            ```

- [ ] Test that it all works and makes sense

3. **Run Tests and Commit Changes**:  
   Make sure your changes don't break existing functionality and commit with a clear message:
   ```bash
   git commit -m "Fix: [Short description of the fix]"
   ```

## Required 📋

To keep our workflow smooth, please make sure you follow these guidelines:

- **Assignment**: Don't create a pull request if you weren’t assigned to this issue.
- **Timeframe**: Complete the task within **3 business days**.
- **Closing the Issue**: In your PR description, close the issue by writing `Close #[issue_id]`.
- **Review Process**:
  - Once you've submitted your PR, change the label to **"ready for review"**.
  - If changes are requested, address them and then update the label back to **"ready for review"** once done.
- **Testing**: Test your PR locally before pushing, and verify that tests and build are working after pushing.

Thank you for your contribution 🙏

⚠️ WARNING: Failure to follow the requirements above may result in being added to the OnlyDust blacklist, affecting your ability to receive future rewards.

### PR DESCRIPTION
**Pull request by abdegenius**
Defi Opportunity Card Component created under the components folder and added at the requirement spot in defiTable.tsx

Closes: #955

## Summary by CodeRabbit

- **New Features**
	- Introduced the `DefiOpportunityCardComponent` to display various liquidity pool and staking opportunities.
	- Enhanced the `DataTable` component with a responsive grid layout showcasing multiple `DefiOpportunityCardComponent` instances.
	- Added utility functions for retrieving token and protocol icons, and parsing token pairs for better representation.
	- Introduced a new number formatting function for improved numerical displays, supporting percentage and currency formats.

- **Bug Fixes**
	- Improved conditional rendering for token icons within the card component.

- **Tests**
	- Added a test suite for the new `formatNumber` function, validating various formatting styles.

### CODE REVIEW

File: components/discover/defiOpportunityCard.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+      </div>
+
+      <div className="w-full grid grid-cols-3 gap-4 my-8 divide-x-[1px] divide-[#29282B]">
+        <div className="w-full mx-auto flex flex-col items-center justify-center text-center">
+          <p className="text-[#A6A5A7] text-[12px] xl:text-[14px] font-[400] mb-1">
+            APR
+          </p>
+          <p className="text-[16px] xl:text-[20px] font-[700] text-[#F4FAFF]">
+            {apr}%
+          </p>
```
[the remark this answers is not in the document]
Comment (fricoben) [reply components/discover/defiOpportunityCard.tsx]: good refactor

File: components/discover/defiOpportunityCard.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      <div className="flex items-center gap-2 w-full">
+        <p className="text-[12px] font-[400] text-[#A6A5A7] w-full">Protocol</p>
+        <div className="flex items-center gap-1.5 flex-none w-auto">
+          <div className="w-[16px] h-[16px] border border-[#A6A5A7] rounded-full overflow-hidden">
+            {/* <img
+              src={protocol.icon}
+              alt={protocol.name}
+              className="w-full h-full object-cover"
+            /> */}
+            <div className="w-full h-full object-over"></div>
```
[the remark this answers is not in the document]
Comment (fricoben) [reply components/discover/defiOpportunityCard.tsx]: Yep !

File: components/discover/defiTable.tsx [hunk 3]
```diff
@@ -413,6 +414,48 @@ const DataTable: FunctionComponent<DataTableProps> = ({ data, loading }) => {
...
+            apr={10}
+            tvl={3.0}
+            dailyRewards={14}
+            protocol={{
+              name: "Nostra Finance",
+              icon: "/icons/nostra-finance.svg",
+            }}
+            token1Icon="/icons/strk.webp"
+            token2Icon=""
+          />
```
[the remark this answers is not in the document]
Comment (fricoben) [reply components/discover/defiTable.tsx]: Don't use hardcoded data! Data must be dynamic based on the filters so you need to use the data we insert in the table.

File: components/discover/defiTable.tsx [hunk 4]
```diff
@@ -413,6 +414,48 @@ const DataTable: FunctionComponent<DataTableProps> = ({ data, loading }) => {
...
+        </div>
+
```
[the remark this answers is not in the document]
Comment (fricoben) [reply components/discover/defiTable.tsx]: You need to integrates in the filtering system as AI says

**Review by fricoben — CHANGES_REQUESTED** [review]
## Code

Lots of code change, please implement

## E2E Verifications

Lots of problems it doesn't look like the figma here:
Capture d’écran 2024-11-25 à 08 54 02
Capture d’écran 2024-11-25 à 08 54 06

- [ ] This should be on one line as said on the figma
Capture d’écran 2024-11-25 à 08 52 54


- [ ] Protocols logo don't work here 
Capture d’écran 2024-11-25 à 08 52 34

- [ ] Protocols icons are undefined, do not work (eth) or aren't high quality enough. Please use [the one we added with this PR](https://github.com/lfglabs-dev/starknet.quest/pull/957)

Capture d’écran 2024-11-25 à 08 54 17

- [ ] Border isn't following figma model

**Comment by abdegenius** [comment]
will look into them all and attend to them properly @fricoben wasn't really clear on some part but this opens up and explains everything,, thanks! will make another PR shortly

**Comment by abdegenius** [comment]
@fricoben please check updates, the icons for token pairs are not on /public/icons and for the protocols am using the favicon.ico file found under each protocol name/subname folder in public(which works good).

File: components/discover/defiTable.tsx [hunk 5]
```diff
@@ -412,6 +446,32 @@ const DataTable: FunctionComponent<DataTableProps> = ({ data, loading }) => {
...
+                  parseTokenPair(opportunity.title).first
+                )}
+                token2Icon={getTokenIcon(
+                  parseTokenPair(opportunity.title).second
+                )}
+              />
+            ))}
+          </div>
+        ) : (
+          ""
```
Comment (fricoben) [inline components/discover/defiTable.tsx]: replace by null

File: components/discover/defiTable.tsx [hunk 6]
```diff
@@ -312,6 +313,39 @@ const DataTable: FunctionComponent<DataTableProps> = ({ data, loading }) => {
...
+        break;
+      } else {
+        [first, second] = [title, ""];
+      }
+    }
+    return {
+      first: first.trim(),
+      second: second?.trim() ?? "",
+    };
+  };
```
[the remark this answers is not in the document]
Comment (fricoben) [reply components/discover/defiTable.tsx]: yes please refactor as AI said

File: components/discover/defiTable.tsx [hunk 7]
```diff
@@ -312,6 +313,15 @@ const DataTable: FunctionComponent<DataTableProps> = ({ data, loading }) => {
...
 
+  const filteredData = table.getRowModel().rows.map((row) => row.original);
+  const topThreeOpportunities = filteredData
+    .sort((a, b) => b.apr - a.apr)
+    .slice(0, 3);
+
+  const getTokenIcon = (token: string): string => {
+    if (!token) return "";
+    return `/icons/${token.toLowerCase()}.svg`;
+  };
```
[the remark this answers is not in the document]
Comment (fricoben) [reply components/discover/defiTable.tsx]: Moreover we should add this to a util file like `defi.ts`

**Review by fricoben — CHANGES_REQUESTED** [review]
Capture d’écran 2024-11-26 à 09 42 29
Capture d’écran 2024-11-26 à 09 42 32

ETH, DAI, WBTC, USDT and STRK don't have a good rendering. Please test that it renders well before sending !

**Comment by abdegenius** [comment]
@fricoben how about the token icons, they are still very much missing from /public/icons.. any insight on how to go about getting them? issue: #954 which is requirement for this current task doesn't seem to have been completed! please advice

**Comment by learnoflix** [comment]
@fricoben still waiting for feedback on this, meanwhile i have pushed the updated code so far

**Comment by abdegenius** [comment]
@fricoben still waiting for your feedback on the icons not found blocker, in the mean time i have pushed the other required updates

**Comment by fricoben** [comment]
@abdegenius Pls updated

**Comment by abdegenius** [comment]
@fricoben please review

**Review by fricoben — CHANGES_REQUESTED** [review]
It doesn't build ! Before asking a review you NEED to:

- [ ] `npm run build` on local
- [ ] Check the remote build after pushing


Capture d’écran 2024-11-28 à 09 48 51

**Comment by abdegenius** [comment]
@fricoben fixed! build works locally and vercel deployment looks good

**Comment by abdegenius** [comment]
@fricoben please review changes..

**Comment by abdegenius** [comment]
@fricoben any update?

**Review by fricoben — CHANGES_REQUESTED** [review]
https://www.loom.com/share/3f47e9c66b8a43c68ffb29740a9895a8

This will be the last review pls check:

- [ ] Use the right icons that are on the testnet branch and look like the figma
- [ ] verify each icon correspond to the right token in the PAIR 
- [ ] verify it builds and works before asking a review

This will be my last review, if it doesn't fit next time I'll finish it myself and it'll affect your reward

**Comment by abdegenius** [comment]
@fricoben recommended changes have been implemented how ever: the icons dimension is say about 60px by 60px but btc and strk even though are of small scale have a lot of padding before the within them created lots of whitespace that make them appear out of resolution, tokens like OPUS,PENDLE, ZKX etc dont have an icon under the tokens folders..

**Comment by abdegenius** [comment]
@fricoben still waiting for your review

**Comment by abdegenius** [comment]
@fricoben please do check the PR!

**Review by fricoben — APPROVED** [review]
lgtm

**Comment by fricoben** [comment]
> @fricoben please do check the PR!

Well please rebase then ...

**Comment by abdegenius** [comment]
@fricoben Rebased! tested everything and the build before pushing.all works good please check

**Comment by abdegenius** [comment]
@fricoben anything else? i have done the rebase

**Pull request merged**